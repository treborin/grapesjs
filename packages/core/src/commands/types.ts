/**{START_EVENTS}*/
export enum CommandsEvents {
  /**
   * @event `command:run` Triggered on run of any command.
   * @example
   * editor.on('command:run', ({ id, result, options }) => {
   *  console.log('Command id', id, 'command result', result);
   * });
   */
  run = 'command:run',

  /**
   * @event `command:run:COMMAND-ID` Triggered on run of a specific command.
   * @example
   * editor.on('command:run:my-command', ({ result, options }) => { ... });
   */
  runCommand = 'command:run:',

  /**
   * @event `command:run:before:COMMAND-ID` Triggered before the command is called.
   * @example
   * editor.on('command:run:before:my-command', ({ options }) => { ... });
   */
  runBeforeCommand = 'command:run:before:',

  /**
   * @event `command:abort:COMMAND-ID` Triggered when the command execution is aborted.
   * @example
   * editor.on('command:abort:my-command', ({ options }) => { ... });
   *
   * // The command could be aborted during the before event
   * editor.on('command:run:before:my-command', ({ options }) => {
   *  if (someCondition) {
   *    options.abort = true;
   *  }
   * });
   */
  abort = 'command:abort:',

  /**
   * @event `command:stop` Triggered on stop of any command.
   * @example
   * editor.on('command:stop', ({ id, result, options }) => {
   *  console.log('Command id', id, 'command result', result);
   * });
   */
  stop = 'command:stop',

  /**
   * @event `command:stop:COMMAND-ID` Triggered on stop of a specific command.
   * @example
   * editor.on('command:run:my-command', ({ result, options }) => { ... });
   */
  stopCommand = 'command:stop:',

  /**
   * @event `command:stop:before:COMMAND-ID` Triggered before the command is called to stop.
   * @example
   * editor.on('command:stop:before:my-command', ({ options }) => { ... });
   */
  stopBeforeCommand = 'command:stop:before:',

  /**
   * @event `command:call` Triggered on run or stop of a command.
   * @example
   * editor.on('command:call', ({ id, result, options, type }) => {
   *  console.log('Command id', id, 'command result', result, 'call type', type);
   * });
   */
  call = 'command:call',

  /**
   * @event `command:call:COMMAND-ID` Triggered on run or stop of a specific command.
   * @example
   * editor.on('command:call:my-command', ({ result, options, type }) => { ... });
   */
  callCommand = 'command:call:',
}
/**{END_EVENTS}*/

// need this to avoid the TS documentation generator to break
export default CommandsEvents;
