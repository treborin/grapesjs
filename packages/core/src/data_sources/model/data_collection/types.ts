import { DataCollectionType, DataCollectionVariableType, keyCollectionDefinition } from './constants';
import { ComponentDefinition, ComponentProperties } from '../../../dom_components/model/types';
import { DataVariableProps } from '../DataVariable';

export type DataCollectionDataSource = DataVariableProps | DataCollectionVariableProps;

export interface DataCollectionConfig {
  collectionId: string;
  startIndex?: number;
  endIndex?: number;
  dataSource: DataCollectionDataSource;
}

export enum DataCollectionStateVariableType {
  currentIndex = 'currentIndex',
  startIndex = 'startIndex',
  currentItem = 'currentItem',
  endIndex = 'endIndex',
  collectionId = 'collectionId',
  totalItems = 'totalItems',
  remainingItems = 'remainingItems',
}

export interface DataCollectionState {
  [DataCollectionStateVariableType.currentIndex]: number;
  [DataCollectionStateVariableType.startIndex]: number;
  [DataCollectionStateVariableType.currentItem]: DataVariableProps;
  [DataCollectionStateVariableType.endIndex]: number;
  [DataCollectionStateVariableType.collectionId]: string;
  [DataCollectionStateVariableType.totalItems]: number;
  [DataCollectionStateVariableType.remainingItems]: number;
}

export interface DataCollectionStateMap {
  [key: string]: DataCollectionState;
}

export interface ComponentDataCollectionProps extends ComponentDefinition {
  [keyCollectionDefinition]: DataCollectionProps;
}

export interface ComponentDataCollectionVariableProps
  extends DataCollectionVariableProps,
    Omit<ComponentProperties, 'type'> {}

export interface DataCollectionProps {
  type: typeof DataCollectionType;
  collectionConfig: DataCollectionConfig;
  componentDef: ComponentDefinition;
}

export interface DataCollectionVariableProps {
  type: typeof DataCollectionVariableType;
  variableType: DataCollectionStateVariableType;
  collectionId: string;
  path?: string;
}
