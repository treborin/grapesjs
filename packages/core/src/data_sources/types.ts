import { Model, Collection, ObjectAny } from '../common';
import DataCollectionVariable from './model/data_collection/DataCollectionVariable';
import { DataCollectionVariableProps } from './model/data_collection/types';
import DataRecord from './model/DataRecord';
import DataRecords from './model/DataRecords';
import DataVariable, { DataVariableProps } from './model/DataVariable';
import { DataConditionProps, DataCondition } from './model/conditional_variables/DataCondition';

export type DataResolver = DataVariable | DataCondition | DataCollectionVariable;

export type DataResolverProps = DataVariableProps | DataConditionProps | DataCollectionVariableProps;

export interface DataRecordProps extends ObjectAny {
  /**
   * Record id.
   */
  id: string;

  /**
   * Specifies if the record is mutable. Defaults to `true`.
   */
  mutable?: boolean;

  [key: string]: any;
}

export interface DataSourceListener {
  obj: Model | Collection;
  event: string;
}

interface BaseDataSource {
  /**
   * DataSource id.
   */
  id: string;

  /**
   * DataSource validation and transformation factories.
   */
  transformers?: DataSourceTransformers;

  /**
   * If true will store the data source in the GrapesJS project.json file.
   */
  skipFromStorage?: boolean;
}
export interface DataSourceType<DR extends DataRecordProps> extends BaseDataSource {
  records: DataRecords<DR>;
}
export interface DataSourceProps<DR extends DataRecordProps> extends BaseDataSource {
  records?: DataRecords<DR> | DataRecord<DR>[] | DR[];
}
export type RecordPropsType<T> = T extends DataRecord<infer U> ? U : never;
export interface DataSourceTransformers {
  onRecordSetValue?: (args: { id: string | number; key: string; value: any }) => any;
}

/**{START_EVENTS}*/
export enum DataSourcesEvents {
  /**
   * @event `data:add` Added new data source.
   * @example
   * editor.on('data:add', (dataSource) => { ... });
   */
  add = 'data:add',
  addBefore = 'data:add:before',

  /**
   * @event `data:remove` Data source removed.
   * @example
   * editor.on('data:remove', (dataSource) => { ... });
   */
  remove = 'data:remove',
  removeBefore = 'data:remove:before',

  /**
   * @event `data:update` Data source updated.
   * @example
   * editor.on('data:update', (dataSource, changes) => { ... });
   */
  update = 'data:update',

  /**
   * @event `data:path` Data record path update.
   * @example
   * editor.on('data:path:SOURCE_ID:RECORD_ID:PROP_NAME', ({ dataSource, dataRecord, path }) => { ... });
   */
  path = 'data:path',

  /**
   * @event `data` Catch-all event for all the events mentioned above.
   * @example
   * editor.on('data', ({ event, model, ... }) => { ... });
   */
  all = 'data',
}
/**{END_EVENTS}*/
type DotSeparatedKeys<T> = T extends object
  ? {
      [K in keyof T]: K extends string
        ? T[K] extends object
          ? `${K}` | `${K}.${DotSeparatedKeys<T[K]>}`
          : `${K}`
        : never;
    }[keyof T]
  : never;

export type DeepPartialDot<T> = {
  [P in DotSeparatedKeys<T>]?: P extends `${infer K}.${infer Rest}`
    ? K extends keyof T
      ? Rest extends DotSeparatedKeys<T[K]>
        ? DeepPartialDot<T[K]>[Rest]
        : never
      : never
    : P extends keyof T
      ? T[P]
      : never;
};
