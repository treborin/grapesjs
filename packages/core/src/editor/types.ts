/**{START_EVENTS}*/
export enum EditorEvents {
  /**
   * @event `update` Event triggered on any change of the project (eg. component added/removed, style changes, etc.)
   * @example
   * editor.on('update', () => { ... });
   */
  update = 'update',

  /**
   * @event `undo` Undo executed.
   * @example
   * editor.on('undo', () => { ... });
   */
  undo = 'undo',

  /**
   * @event `redo` Redo executed.
   * @example
   * editor.on('redo', () => { ... });
   */
  redo = 'redo',

  /**
   * @event `load` Editor is loaded. At this stage, the project is loaded in the editor and elements in the canvas are rendered.
   * @example
   * editor.on('load', () => { ... });
   */
  load = 'load',

  /**
   * @event `project:load` Project JSON loaded in the editor. The event is triggered on the initial load and on the `editor.loadProjectData` method.
   * @example
   * editor.on('project:load', ({ project, initial }) => { ... });
   */
  projectLoad = 'project:load',

  /**
   * @event `project:get` Event triggered on request of the project data. This can be used to extend the project with custom data.
   * @example
   * editor.on('project:get', ({ project }) => { project.myCustomKey = 'value' });
   */
  projectGet = 'project:get',

  /**
   * @event `log` Log message triggered.
   * @example
   * editor.on('log', (msg, opts) => { ... });
   */
  log = 'log',

  /**
   * @event `telemetry:init` Initial telemetry data are sent.
   * @example
   * editor.on('telemetry:init', () => { ... });
   */
  telemetryInit = 'telemetry:init',

  /**
   * @event `destroy` Editor started destroy (on `editor.destroy()`).
   * @example
   * editor.on('destroy', () => { ... });
   */
  destroy = 'destroy',

  /**
   * @event `destroyed` Editor destroyed.
   * @example
   * editor.on('destroyed', () => { ... });
   */
  destroyed = 'destroyed',
}
/**{END_EVENTS}*/
