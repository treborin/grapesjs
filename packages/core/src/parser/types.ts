/**{START_EVENTS}*/
export enum ParserEvents {
  /**
   * @event `parse:html` On HTML parse, an object containing the input and the output of the parser is passed as an argument.
   * @example
   * editor.on('parse:html', ({ input, output }) => { ... });
   */
  html = 'parse:html',
  htmlRoot = 'parse:html:root',

  /**
   * @event `parse:html:before` Event triggered before the HTML parsing starts. An object containing the input is passed as an argument.
   * @example
   * editor.on('parse:html:before', (options) => {
   *   console.log('Parser input', options.input);
   *   // You can also process the input and update `options.input`
   *   options.input += '<div>Extra content</div>';
   * });
   */
  htmlBefore = 'parse:html:before',

  /**
   * @event `parse:css` On CSS parse, an object containing the input and the output of the parser is passed as an argument.
   * @example
   * editor.on('parse:css', ({ input, output }) => { ... });
   */
  css = 'parse:css',

  /**
   * @event `parse:css:before` Event triggered before the CSS parsing starts. An object containing the input is passed as an argument.
   * @example
   * editor.on('parse:css:before', (options) => {
   *   console.log('Parser input', options.input);
   *   // You can also process the input and update `options.input`
   *   options.input += '.my-class { color: red; }';
   * });
   */
  cssBefore = 'parse:css:before',

  /**
   * @event `parse` Catch-all event for all the events mentioned above. An object containing all the available data about the triggered event is passed as an argument to the callback.
   * @example
   * editor.on('parse', ({ event, ... }) => { ... });
   */
  all = 'parse',
}
/**{END_EVENTS}*/
