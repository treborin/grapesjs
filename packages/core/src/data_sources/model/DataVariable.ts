import { Model } from '../../common';
import EditorModel from '../../editor/model/Editor';

export const DataVariableType = 'data-variable' as const;

export interface DataVariableProps {
  type: typeof DataVariableType;
  path: string;
  defaultValue?: string;
}

export default class DataVariable extends Model<DataVariableProps> {
  em?: EditorModel;

  defaults() {
    return {
      type: DataVariableType,
      defaultValue: '',
      path: '',
    };
  }

  constructor(props: DataVariableProps, options: { em?: EditorModel }) {
    super(props, options);
    this.em = options.em;
  }

  getDataValue() {
    const { path, defaultValue } = this.attributes;
    return this.em?.DataSources.getValue(path!, defaultValue);
  }
}
