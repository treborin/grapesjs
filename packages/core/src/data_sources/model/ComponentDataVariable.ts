import Component from '../../dom_components/model/Component';
import { ComponentOptions } from '../../dom_components/model/types';
import { toLowerCase } from '../../utils/mixins';
import DataVariable, { DataVariableProps, DataVariableType } from './DataVariable';

export default class ComponentDataVariable extends Component {
  dataResolver: DataVariable;

  get defaults() {
    return {
      // @ts-ignore
      ...super.defaults,
      type: DataVariableType,
      path: '',
      defaultValue: '',
    };
  }

  constructor(props: DataVariableProps, opt: ComponentOptions) {
    super(props, opt);
    const { type, path, defaultValue } = props;
    this.dataResolver = new DataVariable({ type, path, defaultValue }, opt);
  }

  getDataValue() {
    return this.dataResolver.getDataValue();
  }

  getInnerHTML() {
    return this.getDataValue();
  }

  static isComponent(el: HTMLElement) {
    return toLowerCase(el.tagName) === DataVariableType;
  }
}
