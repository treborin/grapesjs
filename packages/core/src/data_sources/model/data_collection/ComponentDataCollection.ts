import { isArray } from 'underscore';
import { ObjectAny } from '../../../common';
import Component from '../../../dom_components/model/Component';
import { ComponentDefinition, ComponentOptions } from '../../../dom_components/model/types';
import EditorModel from '../../../editor/model/Editor';
import { isObject, serialize, toLowerCase } from '../../../utils/mixins';
import DataResolverListener from '../DataResolverListener';
import DataSource from '../DataSource';
import DataVariable, { DataVariableProps, DataVariableType } from '../DataVariable';
import { isDataVariable } from '../utils';
import { DataCollectionType, keyCollectionDefinition, keyCollectionsStateMap, keyIsCollectionItem } from './constants';
import {
  ComponentDataCollectionProps,
  DataCollectionConfig,
  DataCollectionDataSource,
  DataCollectionProps,
  DataCollectionState,
  DataCollectionStateMap,
} from './types';
import { updateFromWatcher } from '../../../dom_components/model/ComponentDataResolverWatchers';

export default class ComponentDataCollection extends Component {
  constructor(props: ComponentDataCollectionProps, opt: ComponentOptions) {
    const collectionDef = props[keyCollectionDefinition];
    // If we are cloning, leave setting the collection items to the main symbol collection
    if (opt.forCloning) {
      return super(props as any, opt) as unknown as ComponentDataCollection;
    }

    const em = opt.em;
    const newProps = { ...props, components: undefined, droppable: false } as any;
    const cmp: ComponentDataCollection = super(newProps, opt) as unknown as ComponentDataCollection;

    if (!collectionDef) {
      em.logError('missing collection definition');
      return cmp;
    }

    const parentCollectionStateMap = (props[keyCollectionsStateMap] || {}) as DataCollectionStateMap;
    const components: Component[] = getCollectionItems(em, collectionDef, parentCollectionStateMap, opt);
    cmp.components(components, opt);

    if (isDataVariable(this.collectionDataSource)) {
      this.watchDataSource(parentCollectionStateMap, opt);
    }

    return cmp;
  }

  get collectionConfig() {
    return this.get(keyCollectionDefinition).collectionConfig as DataCollectionConfig;
  }

  get collectionDataSource() {
    return this.collectionConfig.dataSource;
  }

  toJSON(opts?: ObjectAny) {
    const json = super.toJSON.call(this, opts) as ComponentDataCollectionProps;
    json[keyCollectionDefinition].componentDef = this.getComponentDef();
    delete json.components;
    delete json.droppable;
    return json;
  }

  private getComponentDef() {
    const firstChild = this.components().at(0);
    const firstChildJSON = firstChild ? serialize(firstChild) : this.get(keyCollectionDefinition).componentDef;
    delete firstChildJSON?.draggable;
    return firstChildJSON;
  }

  private watchDataSource(parentCollectionStateMap: DataCollectionStateMap, opt: ComponentOptions) {
    const { em } = this;
    const path = this.collectionDataSource?.path;
    if (!path) return;

    new DataResolverListener({
      em,
      resolver: new DataVariable({ type: DataVariableType, path }, { em }),
      onUpdate: () => {
        const collectionDef = { ...this.get(keyCollectionDefinition), componentDef: this.getComponentDef() };
        const collectionItems = getCollectionItems(em, collectionDef, parentCollectionStateMap, opt);
        this.components().reset(collectionItems, updateFromWatcher as any);
      },
    });
  }

  static isComponent(el: HTMLElement) {
    return toLowerCase(el.tagName) === DataCollectionType;
  }
}

function getCollectionItems(
  em: EditorModel,
  collectionDef: DataCollectionProps,
  parentCollectionStateMap: DataCollectionStateMap,
  opt: ComponentOptions,
) {
  const { componentDef, collectionConfig } = collectionDef;
  const result = validateCollectionConfig(collectionConfig, componentDef, em);
  if (!result) {
    return [];
  }

  const components: Component[] = [];
  const collectionId = collectionConfig.collectionId;
  const items = getDataSourceItems(collectionConfig.dataSource, em);
  const startIndex = Math.max(0, collectionConfig.startIndex || 0);
  const endIndex = Math.min(
    items.length - 1,
    collectionConfig.endIndex !== undefined ? collectionConfig.endIndex : Number.MAX_VALUE,
  );
  const totalItems = endIndex - startIndex + 1;
  let symbolMain: Component;

  for (let index = startIndex; index <= endIndex; index++) {
    const item = items[index];
    const collectionState: DataCollectionState = {
      collectionId,
      currentIndex: index,
      currentItem: item,
      startIndex: startIndex,
      endIndex: endIndex,
      totalItems: totalItems,
      remainingItems: totalItems - (index + 1),
    };

    if (parentCollectionStateMap[collectionId]) {
      em.logError(
        `The collection ID "${collectionId}" already exists in the parent collection state. Overriding it is not allowed.`,
      );
      return [];
    }

    const collectionsStateMap: DataCollectionStateMap = {
      ...parentCollectionStateMap,
      [collectionId]: collectionState,
    };

    if (index === startIndex) {
      const componentType = (componentDef?.type as string) || 'default';
      let type = em.Components.getType(componentType) || em.Components.getType('default');
      const Model = type.model;
      symbolMain = new Model({ ...serialize(componentDef), draggable: false }, opt);
      setCollectionStateMapAndPropagate(collectionsStateMap, collectionId)(symbolMain);
    }

    const instance = symbolMain!.clone({ symbol: true });
    setCollectionStateMapAndPropagate(collectionsStateMap, collectionId)(instance);

    components.push(instance);
  }

  return components;
}

function setCollectionStateMapAndPropagate(
  collectionsStateMap: DataCollectionStateMap,
  collectionId: string | undefined,
) {
  return (cmp: Component) => {
    setCollectionStateMap(collectionsStateMap)(cmp);

    const addListener = (component: Component) => {
      setCollectionStateMapAndPropagate(collectionsStateMap, collectionId)(component);
    };

    const listenerKey = `_hasAddListener${collectionId ? `_${collectionId}` : ''}`;
    const cmps = cmp.components();

    // Add the 'add' listener if not already in the listeners array
    if (!cmp.collectionStateListeners.includes(listenerKey)) {
      cmp.listenTo(cmps, 'add', addListener);
      cmp.collectionStateListeners.push(listenerKey);

      const removeListener = (component: Component) => {
        component.stopListening(component.components(), 'add', addListener);
        component.off(`change:${keyCollectionsStateMap}`, handleCollectionStateMapChange);
        const index = component.collectionStateListeners.indexOf(listenerKey);
        if (index > -1) {
          component.collectionStateListeners.splice(index, 1);
        }
      };

      cmp.listenTo(cmps, 'remove', removeListener);
    }

    cmps?.toArray().forEach((component: Component) => {
      setCollectionStateMapAndPropagate(collectionsStateMap, collectionId)(component);
    });

    cmp.on(`change:${keyCollectionsStateMap}`, handleCollectionStateMapChange);
  };
}

function handleCollectionStateMapChange(this: Component) {
  const updatedCollectionsStateMap = this.get(keyCollectionsStateMap);
  this.components()
    ?.toArray()
    .forEach((component: Component) => {
      setCollectionStateMap(updatedCollectionsStateMap)(component);
    });
}

function logErrorIfMissing(property: any, propertyPath: string, em: EditorModel) {
  if (!property) {
    em.logError(`The "${propertyPath}" property is required in the collection definition.`);
    return false;
  }
  return true;
}

function validateCollectionConfig(
  collectionConfig: DataCollectionConfig,
  componentDef: ComponentDefinition,
  em: EditorModel,
) {
  const validations = [
    { property: collectionConfig, propertyPath: 'collectionConfig' },
    { property: componentDef, propertyPath: 'componentDef' },
    { property: collectionConfig?.collectionId, propertyPath: 'collectionConfig.collectionId' },
    { property: collectionConfig?.dataSource, propertyPath: 'collectionConfig.dataSource' },
  ];

  for (const { property, propertyPath } of validations) {
    if (!logErrorIfMissing(property, propertyPath, em)) {
      return [];
    }
  }

  return true;
}

function setCollectionStateMap(collectionsStateMap: DataCollectionStateMap) {
  return (cmp: Component) => {
    cmp.set(keyIsCollectionItem, true);
    const updatedCollectionStateMap = {
      ...cmp.get(keyCollectionsStateMap),
      ...collectionsStateMap,
    };
    cmp.set(keyCollectionsStateMap, updatedCollectionStateMap);
    cmp.dataResolverWatchers.updateCollectionStateMap(updatedCollectionStateMap);
  };
}

function getDataSourceItems(dataSource: DataCollectionDataSource, em: EditorModel) {
  let items: DataVariableProps[] = [];

  switch (true) {
    case isArray(dataSource):
      items = dataSource;
      break;
    case isObject(dataSource) && dataSource instanceof DataSource: {
      const id = dataSource.get('id')!;
      items = listDataSourceVariables(id, em);
      break;
    }
    case isDataVariable(dataSource): {
      const isDataSourceId = dataSource.path.split('.').length === 1;
      if (isDataSourceId) {
        const id = dataSource.path;
        items = listDataSourceVariables(id, em);
      } else {
        // Path points to a record in the data source
        items = em.DataSources.getValue(dataSource.path, []);
      }
      break;
    }
    default:
  }

  return items;
}

function listDataSourceVariables(dataSource_id: string, em: EditorModel): DataVariableProps[] {
  const records = em.DataSources.getValue(dataSource_id, []);
  const keys = Object.keys(records);

  return keys.map((key) => ({
    type: DataVariableType,
    path: dataSource_id + '.' + key,
  }));
}
