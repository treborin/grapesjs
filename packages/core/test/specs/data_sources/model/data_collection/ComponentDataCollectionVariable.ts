import { Component, DataRecord, DataSource, DataSourceManager, Editor } from '../../../../../src';
import { DataVariableType } from '../../../../../src/data_sources/model/DataVariable';
import {
  DataCollectionType,
  DataCollectionVariableType,
} from '../../../../../src/data_sources/model/data_collection/constants';
import { DataCollectionStateVariableType } from '../../../../../src/data_sources/model/data_collection/types';
import EditorModel from '../../../../../src/editor/model/Editor';
import { ProjectData } from '../../../../../src/storage_manager';
import { setupTestEditor } from '../../../../common';

describe('Collection variable components', () => {
  let em: EditorModel;
  let editor: Editor;
  let dsm: DataSourceManager;
  let dataSource: DataSource;
  let wrapper: Component;
  let firstRecord: DataRecord;
  let secondRecord: DataRecord;

  beforeEach(() => {
    ({ em, editor, dsm } = setupTestEditor());
    wrapper = em.getWrapper()!;
    dataSource = dsm.add({
      id: 'my_data_source_id',
      records: [
        { id: 'user1', user: 'user1', age: '12' },
        { id: 'user2', user: 'user2', age: '14' },
        { id: 'user3', user: 'user3', age: '16' },
      ],
    });

    firstRecord = dataSource.getRecord('user1')!;
    secondRecord = dataSource.getRecord('user2')!;
  });

  afterEach(() => {
    em.destroy();
  });

  test('Gets the correct static value', async () => {
    const cmp = wrapper.components({
      type: DataCollectionType,
      collectionDef: {
        componentDef: {
          type: 'default',
          components: [
            {
              type: DataCollectionVariableType,
              variableType: DataCollectionStateVariableType.currentItem,
              collectionId: 'my_collection',
              path: 'user',
            },
          ],
        },
        collectionConfig: {
          collectionId: 'my_collection',
          dataSource: {
            type: DataVariableType,
            path: 'my_data_source_id',
          },
        },
      },
    })[0];

    const firstGrandchild = cmp.components().at(0).components().at(0);
    expect(firstGrandchild.getInnerHTML()).toContain('user1');
    expect(firstGrandchild.getEl()?.innerHTML).toContain('user1');

    const secondGrandchild = cmp.components().at(1).components().at(0);
    expect(secondGrandchild.getInnerHTML()).toContain('user2');
    expect(secondGrandchild.getEl()?.innerHTML).toContain('user2');
  });

  test('Watches collection variable changes', async () => {
    const cmp = wrapper.components({
      type: DataCollectionType,
      collectionDef: {
        componentDef: {
          type: 'default',
          components: {
            type: DataCollectionVariableType,
            variableType: DataCollectionStateVariableType.currentItem,
            collectionId: 'my_collection',
            path: 'user',
          },
        },
        collectionConfig: {
          collectionId: 'my_collection',
          dataSource: {
            type: DataVariableType,
            path: 'my_data_source_id',
          },
        },
      },
    })[0];
    firstRecord.set('user', 'new_correct_value');

    const firstGrandchild = cmp.components().at(0).components().at(0);
    expect(firstGrandchild.getInnerHTML()).toContain('new_correct_value');
    expect(firstGrandchild.getEl()?.innerHTML).toContain('new_correct_value');

    const secondGrandchild = cmp.components().at(1).components().at(0);
    expect(secondGrandchild.getInnerHTML()).toContain('user2');
    expect(secondGrandchild.getEl()?.innerHTML).toContain('user2');
  });

  describe('Serialization', () => {
    let cmp: Component;

    beforeEach(() => {
      const variableCmpDef = {
        type: DataCollectionVariableType,
        variableType: DataCollectionStateVariableType.currentItem,
        collectionId: 'my_collection',
        path: 'user',
      };

      const collectionComponentDefinition = {
        type: DataCollectionType,
        collectionDef: {
          componentDef: {
            type: 'default',
            components: [
              {
                type: 'default',
              },
              variableCmpDef,
            ],
          },
          collectionConfig: {
            collectionId: 'my_collection',
            startIndex: 0,
            endIndex: 2,
            dataSource: {
              type: DataVariableType,
              path: 'my_data_source_id',
            },
          },
        },
      };

      cmp = wrapper.components(collectionComponentDefinition)[0];
    });

    test('Serializion to JSON', () => {
      expect(cmp.toJSON()).toMatchSnapshot(`Collection with collection variable component ( no grandchildren )`);

      const firstChild = cmp.components().at(0);
      const newChildDefinition = {
        type: DataCollectionVariableType,
        variableType: DataCollectionStateVariableType.currentIndex,
        collectionId: 'my_collection',
        path: 'user',
      };
      firstChild.components().at(0).components(newChildDefinition);
      expect(cmp.toJSON()).toMatchSnapshot(`Collection with collection variable component ( with grandchildren )`);
    });

    test('Saving', () => {
      const projectData = editor.getProjectData();
      const page = projectData.pages[0];
      const frame = page.frames[0];
      const component = frame.component.components[0];

      expect(component).toMatchSnapshot(`Collection with collection variable component ( no grandchildren )`);

      const firstChild = cmp.components().at(0);
      const newChildDefinition = {
        type: DataCollectionVariableType,
        variableType: DataCollectionStateVariableType.currentIndex,
        collectionId: 'my_collection',
        path: 'user',
      };

      firstChild.components().at(0).components(newChildDefinition);
      expect(cmp.toJSON()).toMatchSnapshot(`Collection with collection variable component ( with grandchildren )`);
    });

    test('Loading', () => {
      const componentProjectData: ProjectData = {
        assets: [],
        pages: [
          {
            frames: [
              {
                component: {
                  components: [
                    {
                      collectionDef: {
                        componentDef: {
                          type: 'default',
                          components: [
                            {
                              type: DataCollectionVariableType,
                              variableType: DataCollectionStateVariableType.currentItem,
                              collectionId: 'my_collection',
                              path: 'user',
                            },
                          ],
                        },
                        collectionConfig: {
                          collectionId: 'my_collection',
                          dataSource: {
                            path: 'my_data_source_id',
                            type: DataVariableType,
                          },
                          endIndex: 1,
                          startIndex: 0,
                        },
                      },
                      type: DataCollectionType,
                    },
                  ],
                  docEl: {
                    tagName: 'html',
                  },
                  head: {
                    type: 'head',
                  },
                  stylable: [
                    'background',
                    'background-color',
                    'background-image',
                    'background-repeat',
                    'background-attachment',
                    'background-position',
                    'background-size',
                  ],
                  type: 'wrapper',
                },
                id: 'frameid',
              },
            ],
            id: 'pageid',
            type: 'main',
          },
        ],
        styles: [],
        symbols: [],
        dataSources: [dataSource],
      };
      editor.loadProjectData(componentProjectData);

      const components = editor.getComponents();
      const component = components.models[0];
      const firstChild = component.components().at(0);
      const firstGrandchild = firstChild.components().at(0);
      const secondChild = component.components().at(1);
      const secondGrandchild = secondChild.components().at(0);

      expect(firstGrandchild.getInnerHTML()).toBe('user1');
      expect(secondGrandchild.getInnerHTML()).toBe('user2');

      firstRecord.set('user', 'new_user1_value');
      expect(firstGrandchild.getInnerHTML()).toBe('new_user1_value');
      expect(secondGrandchild.getInnerHTML()).toBe('user2');

      secondRecord.set('user', 'new_user2_value');
      expect(firstGrandchild.getInnerHTML()).toBe('new_user1_value');
      expect(secondGrandchild.getInnerHTML()).toBe('new_user2_value');
    });
  });
});
