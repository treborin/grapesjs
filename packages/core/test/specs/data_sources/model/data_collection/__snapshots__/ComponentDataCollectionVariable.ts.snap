// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection variable components Serialization Saving: Collection with collection variable component ( no grandchildren ) 1`] = `
{
  "collectionDef": {
    "collectionConfig": {
      "collectionId": "my_collection",
      "dataSource": {
        "path": "my_data_source_id",
        "type": "data-variable",
      },
      "endIndex": 2,
      "startIndex": 0,
    },
    "componentDef": {
      "components": [
        {
          "type": "default",
        },
        {
          "collectionId": "my_collection",
          "path": "user",
          "type": "data-collection-variable",
          "variableType": "currentItem",
        },
      ],
      "type": "default",
    },
  },
  "type": "data-collection",
}
`;

exports[`Collection variable components Serialization Saving: Collection with collection variable component ( with grandchildren ) 1`] = `
{
  "collectionDef": {
    "collectionConfig": {
      "collectionId": "my_collection",
      "dataSource": {
        "path": "my_data_source_id",
        "type": "data-variable",
      },
      "endIndex": 2,
      "startIndex": 0,
    },
    "componentDef": {
      "components": [
        {
          "components": [
            {
              "collectionId": "my_collection",
              "path": "user",
              "type": "data-collection-variable",
              "variableType": "currentIndex",
            },
          ],
          "type": "default",
        },
        {
          "collectionId": "my_collection",
          "path": "user",
          "type": "data-collection-variable",
          "variableType": "currentItem",
        },
      ],
      "type": "default",
    },
  },
  "type": "data-collection",
}
`;

exports[`Collection variable components Serialization Serializion to JSON: Collection with collection variable component ( no grandchildren ) 1`] = `
{
  "collectionDef": {
    "collectionConfig": {
      "collectionId": "my_collection",
      "dataSource": {
        "path": "my_data_source_id",
        "type": "data-variable",
      },
      "endIndex": 2,
      "startIndex": 0,
    },
    "componentDef": {
      "components": [
        {
          "type": "default",
        },
        {
          "collectionId": "my_collection",
          "path": "user",
          "type": "data-collection-variable",
          "variableType": "currentItem",
        },
      ],
      "type": "default",
    },
  },
  "type": "data-collection",
}
`;

exports[`Collection variable components Serialization Serializion to JSON: Collection with collection variable component ( with grandchildren ) 1`] = `
{
  "collectionDef": {
    "collectionConfig": {
      "collectionId": "my_collection",
      "dataSource": {
        "path": "my_data_source_id",
        "type": "data-variable",
      },
      "endIndex": 2,
      "startIndex": 0,
    },
    "componentDef": {
      "components": [
        {
          "components": [
            {
              "collectionId": "my_collection",
              "path": "user",
              "type": "data-collection-variable",
              "variableType": "currentIndex",
            },
          ],
          "type": "default",
        },
        {
          "collectionId": "my_collection",
          "path": "user",
          "type": "data-collection-variable",
          "variableType": "currentItem",
        },
      ],
      "type": "default",
    },
  },
  "type": "data-collection",
}
`;
